ext/ddtrace_profiling_native_extension/collectors_cpu_and_wall_time_worker.c

```diff
@@ -0,0 +1,391 @@
+#include <ruby.h>
+#include <ruby/thread.h>
+#include <ruby/thread_native.h>
+#include <ruby/debug.h>
+#include <stdbool.h>
+#include <signal.h>
+#include "helpers.h"
+#include "ruby_helpers.h"
+#include "collectors_cpu_and_wall_time.h"
+#include "private_vm_api_access.h"
+
+// Used to trigger the periodic execution of Collectors::CpuAndWallTime, which implements all of the sampling logic
+// itself; this class only implements the "doing it periodically" part.
+//
+// This file implements the native bits of the Datadog::Profiling::Collectors::CpuAndWallTimeWorker class
+
+// ---
+// Here be dragons: This component is quite fiddly and probably one of the more complex in the profiler as it deals with
+// multiple threads, signal handlers, global state, etc.
+//
+// ## Design notes for this class:
+//
+// ### Constraints
+//
+// Currently, sampling Ruby threads requires calling Ruby VM APIs that are only safe to call while holding on to the
+// global VM lock (and are not async-signal safe -- cannot be called from a signal handler).
+//
+// @ivoanjo: As a note, I don't we should think of this constraint as set in stone. Since can reach into the Ruby
+// internals, we may be able to figure out a way of overcoming it. But it's definitely going to be hard so for now
+// we're considering it as a given.
+//
+// ### Flow for triggering samples
+//
+// The flow for triggering samples is as follows:
+//
+// 1. Inside the `run_sampling_trigger_loop` function (running in the `CpuAndWallTimeWorker` background thread),
+// a `SIGPROF` signal gets sent to the current process.
+//
+// 2. The `handle_sampling_signal` signal handler function gets called to handle the `SIGPROF` signal.
+//
+//   Which thread the signal handler function gets called on by the operating system is quite important. We need to perform
+// an operation -- calling the `rb_postponed_job_register_one` API -- that can only be called from the thread that
+// is holding on to the global VM lock. So this is the thread we're "hoping" our signal lands on.
+//
+//   The signal never lands on the `CpuAndWallTimeWorker` background thread because we explicitly block it off from that
+// thread in `block_sigprof_signal_handler_from_running_in_current_thread`.
+//
+//   If the signal lands on a thread that is not holding onto the global VM lock, we can't proceed to the next step,
+// and we need to restart the sampling flow from step 1. (There's still quite a few improvements we can make here,
+// but this is the current state of the implementation).
```

Comment:
Are we not able to "detect" what thread is currently evaluating VM code and has the GIL?

Reply:
No...ish.

_Regarding the no part_:

With the current approach no, because that's inherently a racy operation. The signal is asynchronous, so even if you knew the correct thread when you sent the signal, you may have just missed it by the time it gets there.

I think there may be ways to approximate -- we can try asking the kernel which threads are active and send signals to all those OR we can try reading from a couple of in-memory places where Ruby stores pointers to the thread that is supposedly the active one (`ruby_current_execution_context_ptr` / `vm->running_thread` / `rb_thread_set_current()`). But again because everything is moving at the same time, we can still land on the wrong thread because the world changed between our picking and the signal landing.

_Regarding the ish part_:

There's a way of asking the (Linux) kernel to deliver a signal based on cpu-time spent by a thread. Since the thread is the one that spent the time, it's probably the active one (* it may be active but not holding the GVL). But that wouldn't work for wall-clock profiling (e.g. if the app was not very active) so... it wouldn't be a full solution. I may look into using it just for cpu-time profiling at some point.

Reply:
> try reading from a couple of in-memory places where Ruby stores pointers to the thread that is supposedly the active one

I was of thinking of `vm->running_thread` when I wrote that comment, but I understand it's a racy operation with not guarantees.

It seems like a situation where a heuristic would likely be very successfully, despite no solution being 100% robust.

> asking the (Linux) kernel to deliver a signal based on cpu-time spent by a thread.

This seems like a reasonable heuristic, all else failed, btw :)